lm_eval/models/hf_steered.py

```diff
@@ -0,0 +1,254 @@
+
+        [
+            {
+                "hookpoint": <str>,
+                "steering_vector": <torch.Tensor>,
+                "steering_coefficient": <float>,
+                "action": <Literal["add", "clamp_max", "clamp_constant"]>,
+            },
+            ...
+        ]
+        """
+        super().__init__(pretrained=pretrained, device=device, **kwargs)
+
+        if steer_path.endswith(".pt") or steer_path.endswith(".pth"):
+            with open(steer_path, "rb") as f:
+                steer_config = torch.load(f)
+        elif steer_path.endswith(".csv"):
+            steer_config = self.derive_steer_config(steer_path)
+        else:
+            raise ValueError(f"Unknown steer file type: {steer_path}")
+
+        hook_to_steer = {}
+        for hookpoint, steer_info in steer_config.items():
+            action = steer_info["action"]
+            steering_coefficient = steer_info["steering_coefficient"]
+            steering_vector = steer_info["steering_vector"]
+
+            if action == "add":
+                # Steers the model by adding some multiple of a steering vector to all sequence positions.
+                hook_to_steer[hookpoint] = (
+                    lambda acts: acts + steering_coefficient * steering_vector
+                )
+            elif action == "clamp_max":
+                hook_to_steer[hookpoint] = partial(
+                    self.ceiling_clamp,
+                    steering_vector=steering_vector,
+                    value=steering_coefficient,
+                )
+            elif action == "clamp_constant":
+                hook_to_steer[hookpoint] = partial(
+                    self.clamp_constant,
+                    steering_vector=steering_vector,
+                    value=steering_coefficient,
+                )
+            else:
+                raise ValueError(f"Unknown hook type: {action}")
+
+        self.hook_to_steer = hook_to_steer
+
+    @classmethod
+    def derive_steer_config(cls, steer_path: str):
+        """Derive a dictionary of steering vectors from sparse model(/s) specified in a CSV file."""
+        import pandas as pd
+
+        df = pd.read_csv(steer_path)
+        steer_data = {}
+
+        if any(df["loader"] == "sparsify"):
+            from sparsify import SparseCoder
+        if any(df["loader"] == "sae_lens"):
+            from sae_lens import SAE
+
+            sae_cache = {}
+
+            def load_from_sae_lens(sae_release: str, sae_id: str):
+                cache_key = (sae_release, sae_id)
+                if cache_key not in sae_cache:
+                    sae_cache[cache_key] = SAE.from_pretrained(sae_release, sae_id)[0]
+
+                return sae_cache[cache_key]
+
+        for _, row in df.iterrows():
+            action = row.get("action", "add")
+            sparse_name = row["sparse_model"]
+            hookpoint = row["hookpoint"]
```

Comment:
I thought about not requiring this because SAELens has a hookpoint field in their models but for the examples I tried (gemma 2b) the hookpoint it returned seemed to be wrong. TODO double check this